src/App.ts

```diff
@@ -1027,27 +1055,66 @@ export default class App<AppCustomContext extends StringIndexed = StringIndexed>
       // biome-ignore lint/suspicious/noExplicitAny: different kinds of acks accept different arguments, TODO: revisit this to see if we can type better
-      ack?: AckFn<any>;
+      ack: AckFn<any>;
       complete?: FunctionCompleteFn;
       fail?: FunctionFailFn;
       inputs?: FunctionInputs;
     } = {
       body: bodyArg,
+      ack,
       payload,
     };
 
+    // Get the client arg
+    let { client } = this;
+
+    const token = selectToken(context, this.attachFunctionToken);
+
+    if (token !== undefined) {
+      let pool: WebClientPool | undefined = undefined;
+      const clientOptionsCopy = { ...this.clientOptions };
+      if (authorizeResult.teamId !== undefined) {
+        pool = this.clients[authorizeResult.teamId];
+        if (pool === undefined) {
+          pool = this.clients[authorizeResult.teamId] = new WebClientPool();
+        }
+        // Add teamId to clientOptions so it can be automatically added to web-api calls
+        clientOptionsCopy.teamId = authorizeResult.teamId;
+      } else if (authorizeResult.enterpriseId !== undefined) {
+        pool = this.clients[authorizeResult.enterpriseId];
+        if (pool === undefined) {
+          pool = this.clients[authorizeResult.enterpriseId] = new WebClientPool();
+        }
+      }
+      if (pool !== undefined) {
+        client = pool.getOrCreate(token, clientOptionsCopy);
+      }
+    }
+
     // TODO: can we instead use type predicates in these switch cases to allow for narrowing of the body simultaneously? we have isEvent, isView, isShortcut, isAction already in types/utilities / helpers
     // Set aliases
     if (type === IncomingEventType.Event) {
-      const eventListenerArgs = listenerArgs as SlackEventMiddlewareArgs;
+      // TODO: assigning eventListenerArgs by reference to set properties of listenerArgs is error prone, there should be a better way to do this!
+      const eventListenerArgs = listenerArgs as unknown as SlackEventMiddlewareArgs;
       eventListenerArgs.event = eventListenerArgs.payload;
       if (eventListenerArgs.event.type === 'message') {
         const messageEventListenerArgs = eventListenerArgs as SlackEventMiddlewareArgs<'message'>;
         messageEventListenerArgs.message = messageEventListenerArgs.payload;
       }
+      if (eventListenerArgs.event.type === 'function_executed') {
+        listenerArgs.complete = createFunctionComplete(context, client);
+        listenerArgs.fail = createFunctionFail(context, client);
+        listenerArgs.inputs = eventListenerArgs.event.inputs;
+      }
     } else if (type === IncomingEventType.Action) {
       const actionListenerArgs = listenerArgs as SlackActionMiddlewareArgs;
       actionListenerArgs.action = actionListenerArgs.payload;
+      // Add complete() and fail() utilities for function-related interactivity
+      if (context.functionExecutionId !== undefined) {
+        listenerArgs.complete = createFunctionComplete(context, client);
```

Comment:
I think we may need these helpers for View events, too, since those can be function-scoped. [The table in this doc](https://api.slack.com/automation/interactivity#interactivity-handlers) lists out the events that can be function-scoped: actions, view events (view_submission and view_closed) and suggestions aka "options."

Overall there is probably a better way to do these utility-method additions/enrichment differently, but that will be a task for bolt v5, where we leverage the type system to do most of that work for us (rather than fighting it like we do today in this entire `processEvent` method).

Reply:
I was under the impression that Bolt would only support  function-scoped `actions` for the time being and that support for other function-scoped handlers would be added later 🤔 

@misscoded can you confirm this?

Reply:
Correct, I don't believe support for `view` events was introduced/handled at all since interactivity wasn't supported, anyway.

Reply:
Ahhh yes that's right, view events were ignored because of the interactivity-trigger-cross-app permission issue. Got it.
What about block suggestion events?

Reply:
If memory serves, aren't block suggestion events are just actions and handled as such? If not, and they require the same missing piece as `view`, then no, also not supported.

Reply:
Opened [this issue](https://github.com/slackapi/bolt-js/issues/2322) to track this